src/fsharp/ErrorResolutionHints.fs

```diff
@@ -23,60 +56,54 @@ let IsInEditDistanceProximity idText suggestion =
+    let uppercaseText = idText.ToUpperInvariant()
 
-    if allSuggestions.Contains idText then [] else // some other parsing error occurred
     let dotIdText = "." + idText
-    allSuggestions
-    |> Seq.choose (fun suggestion ->
+
+    fun (suggestion:string) ->
         // Because beginning a name with _ is used both to indicate an unused
         // value as well as to formally squelch the associated compiler
         // error/warning (FS1182), we remove such names from the suggestions,
         // both to prevent accidental usages as well as to encourage good taste
-        if IsOperatorName suggestion || suggestion.StartsWithOrdinal("_") then None else
-        let suggestion:string = demangle suggestion
-        let suggestedText = suggestion.ToUpperInvariant()
-        let similarity = EditDistance.JaroWinklerDistance uppercaseText suggestedText
-        if similarity >= highConfidenceThreshold || suggestion.EndsWithOrdinal(dotIdText) then
-            Some(similarity, suggestion)
-        elif similarity < minThresholdForSuggestions && suggestedText.Length > minStringLengthForThreshold then
-            None
-        elif IsInEditDistanceProximity uppercaseText suggestedText then
-            Some(similarity, suggestion)
+        if IsOperatorName suggestion || suggestion.StartsWithOrdinal("_") then 
+            ()
         else
-            None)
-    |> Seq.sortByDescending fst
-    |> Seq.mapi (fun i x -> i, x) 
-    |> Seq.takeWhile (fun (i, _) -> i < maxSuggestions) 
-    |> Seq.map snd 
-    |> Seq.toList
+            let suggestion:string = DemangleSuggestion suggestion
+            let suggestedText = suggestion.ToUpperInvariant()
+            let similarity = EditDistance.JaroWinklerDistance uppercaseText suggestedText
+            if similarity >= highConfidenceThreshold || suggestion.EndsWithOrdinal(dotIdText) then
+                buffer.Insert(similarity, suggestion) |> ignore
+            elif similarity < minThresholdForSuggestions && suggestedText.Length > minStringLengthForThreshold then
+                ()
+            elif IsInEditDistanceProximity uppercaseText suggestedText then
+                buffer.Insert(similarity, suggestion) |> ignore
 
-/// Formats the given predictions according to the error style.
-let FormatPredictions normalizeF (predictions: (float * string) list) =
-    match predictions with
-    | [] -> System.String.Empty
-    | _ ->
+/// Formats the given suggestions according to the error style.
+let FormatSuggestions normalizeF (idText:string) (buffer: ErrorLogger.MaxBuffer<_,string>) =
+    let values = buffer.Values()
+    if Array.exists ((=) idText) values then
+        System.String.Empty
+    else
         let suggestions =
-            predictions
-            |> List.map (snd >> normalizeF) 
-            |> List.map (sprintf "%s   %s" System.Environment.NewLine)
+            values
+            |> Seq.map (fun p -> sprintf "%s   %s" System.Environment.NewLine (normalizeF p))
```

Comment:
We should have `FormatSuggestions` accept a `StringBuilder` instance so that we can then use a for loop over IEnumerable to just `AddFormat` on `StringBuilder` rather then using `String.Concat` ? I think the caller of  `FormatSuggestions` has `StringBuilder` already in scope?

Reply:
Yep moving the SB down is a good improvement

Reply:
done